### PR TITLE
Add an e2e test to our pipeline

### PR DESCRIPTION
**Pull request by marbre**
This finally publishes and utilized our internal script to generate and run an e2e test as part of our GitHub action.

### CODE REVIEW

**Comment by marbre** [comment]
@simon-camp please review the proposed changes and I will rebase and merge after your approval. I think about moving parts of the stuff now hacked into the action itself to a separate bash script, similar to the one you wrote for our internal use. Anyway, this should be handled as a followup.

File: .github/workflows/build-and-test.yaml [hunk 1]
```diff
@@ -65,12 +65,46 @@ jobs:
...
+    - name: Cache e2e
+      id: cache-e2e
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.E2E }}
+        key: ${{ runner.os }}-e2e-${{ hashFiles('emitc/scripts/*.py') }}
+
+    - name: Install TensorFlow
+      if: steps.cache-e2e.outputs.cache-hit != 'true'
+      run: python3 -m pip install tensorflow==2.4
```
Comment (simon-camp) [inline .github/workflows/build-and-test.yaml:87]: We can move the dependencies to a requirements.txt file. This makes it easier to reproduce for users.

File: .github/workflows/build-and-test.yaml [hunk 2]
```diff
@@ -124,3 +160,21 @@ jobs:
...
+        path: ${{ env.E2E }}
+        key: ${{ runner.os }}-e2e-${{ hashFiles('emitc/scripts/*.py') }}
+
+    - name: Run e2e test
+      run: |
+        ./${EMITC}/build_release/bin/emitc-opt --canonicalize --inline --symbol-dce ${E2E}/model_mhlo.mlir > ${E2E}/model_canon.mlir
+        FUNCTION_NAME=$(grep -oe "@[^(]*" "${E2E}"/model_canon.mlir)
+        FUNCTION_NAME="${FUNCTION_NAME:1}"
+        sed "s/$FUNCTION_NAME/predict/g" "${E2E}"/model_canon.mlir > "${E2E}"/model_fix_name.mlir
+        ./${EMITC}/build_release/bin/emitc-opt --mhlo-control-flow-to-scf -convert-mhlo-region-ops-to-emitc --convert-mhlo-to-emitc --convert-scf-to-emitc --convert-std-to-emitc --convert-mhlo-const-to-std "${E2E}"/model_fix_name.mlir > "${E2E}"/model_emitc.mlir
```
Comment (simon-camp) [inline .github/workflows/build-and-test.yaml:176]: nit. Be consistent with - and --
```suggestion
        ./${EMITC}/build_release/bin/emitc-opt --mhlo-control-flow-to-scf --convert-mhlo-region-ops-to-emitc --convert-mhlo-to-emitc --convert-scf-to-emitc --convert-std-to-emitc --convert-mhlo-const-to-std "${E2E}"/model_fix_name.mlir > "${E2E}"/model_emitc.mlir
```

**Review by simon-camp — APPROVED** [review]

File: .github/workflows/build-and-test.yaml (hunk 1 above)
Comment (marbre) [reply .github/workflows/build-and-test.yaml]: Yeah, I intend to add some documentation on our e2e flow anyway, but this is a good point.

**Pull request merged**